native_c_extension/c_custom_to_s.c

```diff
@@ -0,0 +1,493 @@
+#include "ruby.h"
+
+// Defining a space for information and references about the module to be stored internally
+VALUE CCustomToS = Qnil;
+
+// Prototype for the initialization method - Ruby calls this, not you
+void Init_ccustomtos();
+
+// Prototype for our methods, to be used by Init_ccustomtos
+VALUE method_c_custom_to_s(VALUE, VALUE);
+
+// The initialization method for this module
+void Init_ccustomtos() {
+  CCustomToS = rb_define_module("CCustomToS");
+  rb_define_method(CCustomToS, "c_custom_to_s", method_c_custom_to_s, 1);
+}
+
+const int MAX_ARRAY_ENUMERATION = 10;
+const int MAX_HASH_ENUMERATION = 10;
+const int MAX_STRING_LENGTH = 100;
+const int MAX_EXPECTED_NESTED_LEVELS = 4;
+
+void method_c_custom_to_s_check_buffer_size(int offset, int string_len, int buffer_max) {
+  if (offset + string_len > buffer_max) {
+    // don't corrupt the buffer; throw exception
+    rb_raise(rb_eRuntimeError, "method_c_custom_t_s_element would write %d bytes outside the buffer", offset + string_len - buffer_max);
+  }
+}
+
+int method_c_custom_to_s_element(VALUE self, char *buffer, int *offset, VALUE element, int buffer_max, char quoted) {
+  switch (TYPE(element)) {
+  case T_NIL: {
+    int string_len = 3; // +1 for NULL
+    method_c_custom_to_s_check_buffer_size(*offset, string_len + 1, buffer_max);
+    snprintf(&buffer[*offset], string_len + 1, "%s", "nil");
+    *offset += string_len;
+    break;
+  }
+  case T_STRING: {
+    VALUE element_to_s = method_c_custom_to_s(self, element);
+    int string_len = RSTRING_LEN(element_to_s);
+    // +2 for the two "s. +1 for NULL
+    method_c_custom_to_s_check_buffer_size(*offset, string_len + 3, buffer_max);
+    if (quoted) {
+      snprintf(&buffer[*offset], 1 + 1, "%s", "\"");
+      *offset += 1;
+    }
+    // +1 for NULL
+    snprintf(&buffer[*offset], string_len + 1, "%s", StringValuePtr(element_to_s));
```

Comment:
Again, snprintf is unnecessary here. What's worse, I think it can do the wrong thing -- ruby strings aren't guaranteed to be null-terminated. You need to do something like
```c
memcpy(buffer + *offset, StringValuePtr(element_to_s), string_len);
*offset += string_len;
```

Reply:
Switched snprintf to memcpy.